### PR TITLE
modules: support tests/samples/boards in modules

### PR DESCRIPTION
**Pull request by nashif**
Generate options for sanitycheck to run tests and samples in modules.
Use the --sanitycheck-out <file> to generate a file that can be supplied
to sanitycheck on the commandline which will add additional testroots
and boards if the module does contain out of tree boards.

the module.yaml file now accepts the following:

samples:
  path: samples
tests:
  path: tests
boards:
  path: boards

- [x] Finalize `module.yml` structure
- [x] Documentation needed
- [ ] Test/sample module needed

### CODE REVIEW

File: scripts/zephyr_module.py [hunk 1]
```diff
@@ -71,44 +89,79 @@ def process_module(module, cmake_out=None, kconfig_out=None):
...
+def main():
     parser = argparse.ArgumentParser(description='''
     Process a list of projects and create Kconfig / CMake include files for
     projects which are also a Zephyr module''')
 
     parser.add_argument('--kconfig-out',
                         help='File to write with resulting KConfig import'
                              'statements.')
+    parser.add_argument('--sanitycheck-out',
+                        help='File to write with resulting sanitycheck'
```
Comment (mbolivar) [inline scripts/zephyr_module.py]: I think you're just copy pasting the style from other arguments, but there is a missing space.  It should be `sanitycheck '` instead.

**Review by mbolivar — COMMENTED** [review]
Appreciate this! 

Can we get some corresponding documentation changes before merging this?

**Review by carlescufi — DISMISSED** [review]
Thanks for the PR!
Can you add the relevant changes to:
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/zephyr_module.cmake#L30
as part of this PR?

**Comment by carlescufi** [comment]
> Thanks for the PR!
> Can you add the relevant changes to:
> https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/zephyr_module.cmake#L30
> as part of this PR?

Actually, my mistake, `zephyr_module.cmake` is only for building samples, so I assume you won't do that here. But where are you going to invoke `zephyr_module.py --sanitycheck-out` ?

File: scripts/zephyr_module.py [hunk 2]
```diff
@@ -39,6 +39,27 @@
...
+        type: str
+        required: false
+  samples:
+    required: false
+    type: map
+    mapping:
+      path:
+        type: str
+        required: false
+  boards:
```
Comment (carlescufi) [inline scripts/zephyr_module.py]: Why did you add boards here, since that should really be part of the `sample.yaml` or `test.yaml`?

**Comment by nashif** [comment]
> But where are you going to invoke `zephyr_module.py --sanitycheck-out` ?

in CI, generate the file and let sanitycheck loose on samples/tests in modules.

File: doc/guides/modules.rst [hunk 3]
```diff
@@ -82,6 +82,26 @@ Each project in the ``west list`` output is tested like this:
...
+  .. code-block:: yaml
+
+      build:
+        cmake: .
+      samples:
+        path: samples
+      tests:
+        path: tests
+      boards:
+        path: boards
```
Comment (tejlmand) [inline doc/guides/modules.rst]: Now, the existing configuration is actually implying a path:
`cmake: <path>` and `kconfig: <path/Kconfig-file>` under the build, so I kind of feel strange that `path` is a key by itself.

How about:
```suggestion
      samples: <path/to/samples>
      tests: <path/to/tests>
      boards: <path/to/boards>
```
where the path is relative from beginning of the module, which would most often be:
```
      samples: samples
      tests: tests
      boards: boards
```

Of course we might want an additional level besides `build`, as example `source` ? (feel free to provide better name)

Comment (carlescufi) [reply doc/guides/modules.rst]: @tejlmand 
That's because you can have multiple `samples:` and `tests:` but you can't have multiple CMake or Kconfig files.

**Review by carlescufi — CHANGES_REQUESTED** [review]
You need this patch to make this work with MCUboot:

```
diff --git a/scripts/zephyr_module.py b/scripts/zephyr_module.py
index 1d15a87743..eeec936f23 100755
--- a/scripts/zephyr_module.py
+++ b/scripts/zephyr_module.py
@@ -35,7 +35,7 @@ METADATA_SCHEMA = '''
 type: map
 mapping:
   build:
-    required: true
+    required: false
     type: map
```

Other than that it seems to work fine, thank you!

**Comment by carlescufi** [comment]
@mbolivar @tejlmand can you take another look and see if you are happy with the proposal?

File: doc/guides/modules.rst (hunk 3 above)
Comment (carlescufi) [reply doc/guides/modules.rst]: @tejlmand @nashif @mbolivar my proposal would be to accept multiple entries for `path:`:

```
samples:
  path: path/to/sample/one
  paht: path/to/sample/two
```

Comment (carlescufi) [reply doc/guides/modules.rst]: Or, as @nashif mentioned perhaps:

```
samples:
  - path/to/sample/one
  - path/to/sammple/two
```

Comment (mbolivar) [reply doc/guides/modules.rst]: Doing it as a list is the right way YAML-wise.

**Review by carlescufi — APPROVED** [review]
Tested alongside:
https://github.com/zephyrproject-rtos/zephyr/pull/17868
and
https://github.com/JuulLabs-OSS/mcuboot/pull/619

File: doc/guides/modules.rst [hunk 4]
```diff
@@ -82,6 +82,26 @@ Each project in the ``west list`` output is tested like this:
        cmake: .
        kconfig: Kconfig
 
+- To execute both tests and samples available in modules the Zephyr test runner
```
Comment (mbolivar) [inline doc/guides/modules.rst]: s/samples in tests/samples and tests/

File: scripts/zephyr_module.py [hunk 5]
```diff
@@ -10,6 +10,11 @@
 
 Include file is generated for Kconfig using --kconfig-out.
 A <name>:<path> text file is generated for use with CMake using --cmake-out.
+
+Using --sanitycheck-out <filenam> an argument file for sanitycheck script will
```
Comment (mbolivar) [inline scripts/zephyr_module.py]: filename

File: doc/guides/modules.rst [hunk 6]
```diff
@@ -82,6 +82,26 @@ Each project in the ``west list`` output is tested like this:
...
+        cmake: .
+      samples:
+        - samples
+      tests:
+        - tests
+      boards:
+        - boards
+
+  Additionally, if a module defines out of tree boards, the module file can
+  point sanitycheck to the path where those files are maintained in the module.
```
Comment (mbolivar) [inline doc/guides/modules.rst]: I think this ought to be before the code sample, no? It would be clearer.

File: scripts/zephyr_module.py [hunk 7]
```diff
@@ -10,6 +10,11 @@
...
+be generated which would point to test and sample roots available in modules
+that can be included during a sanitycheck run. This allows testing code
+maintaiend in modules in addition to what is available in the main Zephyr tree.
```
Comment (mbolivar) [inline scripts/zephyr_module.py]: Maintained

**Review by mbolivar — CHANGES_REQUESTED** [review]
Minor tweaks to docs suggested.

I wish this could be made simpler, to be honest, but I'm in favor of getting the feature in and iterating as needed.

**Review by mbolivar — APPROVED** [review]
